src/node_crypto.cc

```diff
@@ -240,6 +240,43 @@ static int PasswordCallback(char *buf, int size, int rwflag, void *u) {
   return 0;
 }
 
+// Loads OpenSSL engine by engine id and returns it. The loaded engine
+// gets a reference so remember the corresponding call to ENGINE_free.
+// In case of error the appropriate js exception is scheduled
+// and nullptr is returned.
+#ifndef OPENSSL_NO_ENGINE
+static ENGINE* LoadEngineById(Environment* env, const char* engine_id) {
+  ENGINE* engine = ENGINE_by_id(engine_id);
+
+  MarkPopErrorOnReturn mark_pop_error_on_return;
+
+  if (engine == nullptr) {
+    // Engine not found, try loading dynamically.
+    engine = ENGINE_by_id("dynamic");
+    if (engine != nullptr) {
+      if (!ENGINE_ctrl_cmd_string(engine, "SO_PATH", engine_id, 0) ||
+          !ENGINE_ctrl_cmd_string(engine, "LOAD", nullptr, 0)) {
+        ENGINE_free(engine);
+        engine = nullptr;
+      }
+    }
+  }
+
+  if (engine == nullptr) {
+    char errmsg[1024];
+    int err = ERR_get_error();
+    if (err != 0) {
+      ERR_error_string_n(err, errmsg, sizeof(errmsg));
+    } else {
+      snprintf(errmsg, sizeof(errmsg),
+               "Engine \"%s\" was not found", engine_id);
+    }
+    env->ThrowError(errmsg);
```

Comment:
I mentioned it in the other PR but throwing a JS exception here is kind of meh.  If you change the function prototype to this:
```c++
static ENGINE* LoadEngineById(const char* engine_id, char (*errmsg)[1024]) {
```
Then you can `snprintf(*errmsg, sizeof(*errmsg), ...)` here.  At the call site:
```c++
char errmsg[1024];
if (auto engine = LoadEngineById(id, &errmsg))
  // do something
else
  return env->ThrowError(errmsg);
```
That also gets rid of the otherwise unnecessary dependency on `Environment`.